### PR TITLE
Add login v2. Stop serving the environment at the root of the API.

### PR DESCRIPTION
**Pull request by howbazaar**
This change set has a bunch of mostly related changes.

I added a ServerVersion to the login response. This is set for the version two login.

The creation of some machines and units in tests were simplified to use the factory.

The api package gains a new exported method: OpenWithVersion that allows specifying the version of the Login to use. This is only used in tests and allows the client to pretend to be older.

### CODE REVIEW

File: api/apiclient.go [hunk 1]
```diff
@@ -166,6 +180,21 @@ func Open(info *Info, opts DialOpts) (*State, error) {
 	return st, nil
 }
 
+// OpenWithVersion uses an explicit version of the Admin facade to call Login
+// on.  This allows the caller to pretends to be an older client, and is used
+// only in testing.
+func OpenWithVersion(info *Info, opts DialOpts, loginVersion int) (*State, error) {
+	switch loginVersion {
+	case 0:
+		return open(info, opts, (*State).loginV0)
```
Comment (mjs) [inline api/apiclient.go]: Nit: given that the case blocks are all the same except for the login method arg, I'd prefer to see the login method assigned to a variable in each case block, with `open` called just once at the bottom of the func.

Not a biggie though.

File: api/apiclient.go [hunk 2]
```diff
@@ -136,6 +142,14 @@ func DefaultDialOpts() DialOpts {
 //
 // See Connect for details of the connection mechanics.
 func Open(info *Info, opts DialOpts) (*State, error) {
+	return open(info, opts, (*State).Login)
+}
+
+// This unexported open method is used both directly above in the Open
```
Comment (mjs) [inline api/apiclient.go]: This comment doesn't line up with how things ended up.

File: api/client.go [hunk 3]
```diff
@@ -987,10 +987,22 @@ func (c *Client) WatchDebugLog(args DebugLogParams) (io.ReadCloser, error) {
...
+		// serving debug log at the environment path. We also fully expect
+		// that the server has returned a valid environment tag.
+		envTag, err := c.st.EnvironTag()
+		if err != nil {
+			return nil, errors.Annotate(err, "very unexpected")
+		}
+		path = fmt.Sprintf("/environment/%s/log", envTag.Id())
+	}
+
 	target := url.URL{
```
Comment (mjs) [inline api/client.go]: Just noticed that the following bit can all be replaced with a call to `api.Connect`. It's not for this PR though.

File: apiserver/restricted_root.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+type restrictedRoot struct {
+	rpc.MethodFinder
+}
+
+// newRestrictedRoot returns a new restrictedRoot.
+func newRestrictedRoot(finder rpc.MethodFinder) *restrictedRoot {
+	return &restrictedRoot{finder}
+}
+
+var restrictedRootNames = set.NewStrings(
```
Comment (mjs) [inline apiserver/restricted_root.go]: A docstring here would be nice just to clarify what these names are

File: apiserver/restricted_root.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"EnvironmentManager",
+	"UserManager",
+)
+
+// FindMethod returns a not supported error if the rootName is not one
+// of the facades available at the server root when there is no active
+// environment.
+func (r *restrictedRoot) FindMethod(rootName string, version int, methodName string) (rpcreflect.MethodCaller, error) {
+	if !restrictedRootNames.Contains(rootName) {
```
Comment (mjs) [inline apiserver/restricted_root.go]: The problem with this approach is that for completely invalid root names the client gets the same error as if they tried to access a root that's no longer available.

The other roots (e.g. upgradingRoot) do FindMethod first and let that generate a specific error for completely wrong calls, and *then* filter. Consider doing the same thing here.

File: doc/logging-in-to-the-apiserver.txt [hunk 6]
```diff
@@ -0,0 +1,92 @@
+Logging in to the API Server
```
Comment (mjs) [inline doc/logging-in-to-the-apiserver.txt]: This is great. Thanks.

File: doc/logging-in-to-the-apiserver.txt [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+In order to get the envuuid from the path into the Login method, we need to
+capture it in *Server.serveConn, where it is passed in as an argument.  We
+should pass it through to the newApiHandler function and store it on the
+apiHandler struct.  This handler is passed through as an arg to the
+newAnonRoot, and to the factory methods for the admin APIs.
+
+
+Implementation Notes:
+
+The login process for both v0 and v1 logins are handled in the method *admin.doLogin (in admin.go).  The method *Conn.ServeFinder on the *rpc.Conn attribute of the apiHandler is initially given the anonymous root here:
```
Comment (mjs) [inline doc/logging-in-to-the-apiserver.txt]: Line lengths get longer from here on. Make them consistent with the rest of the doc?

**Comment by mjs** [comment]
My comments are all fairly minor. LGTM.

File: api/apiclient.go (hunk 1 above)
Comment (howbazaar) [inline api/apiclient.go]: Done.

File: api/apiclient.go (hunk 2 above)
Comment (howbazaar) [inline api/apiclient.go]: Fixed.

File: apiserver/restricted_root.go (hunk 4 above)
Comment (howbazaar) [inline apiserver/restricted_root.go]: Done.

File: apiserver/restricted_root.go (hunk 5 above)
Comment (howbazaar) [inline apiserver/restricted_root.go]: Done.

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$my-fault$$

**Comment by howbazaar** [comment]
$$replicaset-failure$$

**Pull request merged**